src/coprocessor/dag/dag.rs

```diff
@@ -26,52 +25,55 @@ use coprocessor::{Error, Result};
 use coprocessor::endpoint::{get_chunk, get_pk, to_pb_error, ReqContext};
 use storage::{Snapshot, SnapshotStore, Statistics};
 
-use super::executor::{AggregationExecutor, Executor as DAGExecutor, IndexScanExecutor,
-                      LimitExecutor, Row, SelectionExecutor, TableScanExecutor, TopNExecutor};
+use super::executor::{build_exec, Executor, Row};
 
-pub struct DAGContext<'s> {
-    columns: Rc<Vec<ColumnInfo>>,
+pub struct DAGContext {
+    columns: Arc<Vec<ColumnInfo>>,
     has_aggr: bool,
-    req: DAGRequest,
-    ranges: Vec<KeyRange>,
-    snap: &'s Snapshot,
-    eval_ctx: Rc<EvalContext>,
-    req_ctx: &'s ReqContext,
+    req_ctx: Arc<ReqContext>,
+    exec: Box<Executor>,
+    output_offsets: Vec<u32>,
 }
 
-impl<'s> DAGContext<'s> {
+impl DAGContext {
     pub fn new(
-        req: DAGRequest,
+        mut req: DAGRequest,
         ranges: Vec<KeyRange>,
-        snap: &'s Snapshot,
-        eval_ctx: Rc<EvalContext>,
-        req_ctx: &'s ReqContext,
-    ) -> DAGContext<'s> {
-        DAGContext {
-            req: req,
-            columns: Rc::new(vec![]),
-            ranges: ranges,
-            snap: snap,
-            has_aggr: false,
-            eval_ctx: eval_ctx,
+        snap: Box<Snapshot>,
+        req_ctx: Arc<ReqContext>,
+    ) -> Result<DAGContext> {
+        let eval_ctx = Arc::new(box_try!(EvalContext::new(
+            req.get_time_zone_offset(),
+            req.get_flags()
+        )));
+        let store = SnapshotStore::new(
+            snap,
+            req.get_start_ts(),
+            req_ctx.isolation_level,
+            req_ctx.fill_cache,
+        );
+
+        let dag_executor = build_exec(req.take_executors().into_vec(), store, ranges, eval_ctx)?;
+        Ok(DAGContext {
+            columns: dag_executor.columns,
+            has_aggr: dag_executor.has_aggr,
             req_ctx: req_ctx,
-        }
+            exec: dag_executor.exec,
+            output_offsets: req.take_output_offsets(),
+        })
     }
 
-    pub fn handle_request(mut self, statistics: &'s mut Statistics) -> Result<Response> {
-        self.validate_dag()?;
-        let mut exec = self.build_dag(statistics)?;
+    pub fn handle_request(&mut self) -> Result<Response> {
         let mut chunks = vec![];
         loop {
-            match exec.next() {
+            match self.exec.next() {
                 Ok(Some(row)) => {
                     self.req_ctx.check_if_outdated()?;
                     let chunk = get_chunk(&mut chunks);
```

Comment:
It seems there would be only one chunk since `chunk.get_rows_meta()` would always be empty?